src/modules/RA/Picture/components/CustomToolbar.js

```diff
@@ -65,6 +65,130 @@ const CancelButton = ({ redirect, className }) => (
+  saving,
+  invalid,
+  transform = () => { },
+  ...props
+}) => {
+  const { setOnSuccess, setOnFailure, setTransform } = useSaveContext(props);
+  const translate = useTranslate();
+  const [open, setOpen] = React.useState(false);
+  const [reason, setReason] = React.useState('');
+  const notify = useNotify();
+
+  const handleClick = React.useCallback(e => {
+    if (invalid) {
+      notify('ra.message.invalid_form', 'warning');
+    }
+    if (saving) {
+      // prevent double submission
+      e.preventDefault();
+    }
+    setOpen(true);
+  }, [invalid, notify, saving]);
+
+
+  // Not memoized on purpose
+  const handleConfirm = () => {
+    if (onSuccess) {
+      setOnSuccess(onSuccess);
+    }
+    if (onFailure) {
+      setOnFailure(onFailure);
+    }
+
+    const handleTransform = data => {
+      const tdata = transform(data);
+      return { ...tdata, properties: { ...tdata.properties, refusal_message: reason } };
+    };
+
+    setTransform(handleTransform);
+    handleSubmitWithRedirect(redirect);
+  };
+
+  return (
+    <>
+      <SaveButton {...props} handleSubmitWithRedirect={handleClick} saving={saving} />
+      <Confirm
+        isOpen={open}
+        title={translate('resources.picture.fields.refusal_message')}
+        content={(
+          <>
+            <p>{translate('resources.picture.actions.refusal_modal_content')}</p>
+            <input value={reason} onChange={e => setReason(e.target.value)} />
+          </>
+        )}
+        translateOptions={{
+          id: record.id,
+        }}
+        onConfirm={handleConfirm}
+        onClose={() => setOpen(false)}
+      />
+    </>
+  );
+};
+
+const SaveButtonWithConfirm = ({
+  handleSubmitWithRedirect,
+  record,
+  resource,
+  redirect,
+  onSuccess,
+  onFailure,
+  saving,
+  invalid,
+  transform,
+  ...props
+}) => {
+  const translate = useTranslate();
+  const [open, setOpen] = React.useState(false);
+  const notify = useNotify();
+  const { setOnSuccess, setOnFailure, setTransform } = useSaveContext(props);
+
+  const handleClick = React.useCallback(e => {
+    if (invalid) {
+      notify('ra.message.invalid_form', 'warning');
+    }
+    if (saving) {
+      // prevent double submission
+      e.preventDefault();
+    }
+    setOpen(true);
+  }, [invalid, notify, saving]);
+
+  // Not memoized on purpose
+  const handleConfirm = () => {
+    if (onSuccess) {
+      setOnSuccess(onSuccess);
+    }
+    if (onFailure) {
+      setOnFailure(onFailure);
+    }
+    setTransform(transform);
+    handleSubmitWithRedirect(redirect);
+  };
+
+  return (
+    <>
+      <SaveButton {...props} handleSubmitWithRedirect={handleClick} saving={saving} />
+      <Confirm
+        isOpen={open}
+        title={translate('resources.picture.actions.accept_modal_title')}
+        content={translate('resources.picture.actions.accept_modal_content')}
+        onConfirm={handleConfirm}
+        onClose={() => setOpen(false)}
+      />
+    </>
+  );
+};
+
```

Comment:
There are a lot of duplication for the same work.
Why don't you pass a boolean prop to display the reason or not?

Reply:
Refactored